Wordpress improvements

mohe2015 opened the pull request:
#####  Motivation for this change
Improve the wordpress service:
* Add separate package that removes built-in themes and plugins as these can't be updated independently and also can't be removed otherwise.
* Package auto-update script
* Separate databases to allow multiple wordpress installations on one host by default
* Add option for mutable wp-content

(edit: updated)

###### Things done


- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mohe2015 committed:
wordpress: Add update script

mohe2015 committed:
wordpress-core: init at 5.7.2

mohe2015 committed:
nixos/mysql: Fix for databases with special characters

aanderse commented on nixos/modules/services/web-servers/apache-httpd/default.nix:
```diff
@@ -647,14 +647,15 @@ in
-      extraDomains = genAttrs hostOpts.serverAliases (alias: null);
-      postRun = "systemctl reload httpd.service";
-    }) (filterAttrs (name: hostOpts: hostOpts.enableACME) cfg.virtualHosts);
+    security.acme.certs = mapAttrs' (name: hostOpts: (
```
Good catch. The previous code here assumed the attribute name was the host name which is not always the case! Can you file a separate PR for this change? I'd like to get this fix into `master` as soon as possible as we have feature freeze coming up soon.

ajs124 commented:
I haven't gotten a chance to look at this, but wordpress can actually be served mostly read-only, from the nix store. One component that comes in handy for that is [this ~abomination~](https://git.helsinki.tools/helsinki-systems/wp4nix)beautiful piece of technology. It gives you each and every wordpress theme and plugin as a nix derivation.

mohe2015 commented:
@ajs124 Would it be possible for you to review my PR? I would really appreciate it.
Also regarding [wp4nix](https://git.helsinki.tools/helsinki-systems/wp4nix) I'm only using a few themes and plugins so I think my approach may also be fine (I think yours downloads all themes and plugins doesn't it?). I have to work on the translations in the future though.

ajs124 commented:
Sure, I'll try and give some useful feedback.
But first, wp4nix works by crawling the WordPress plugin directory API and SVN. It outputs a bunch of JSON files, in total roughly 25MB. These JSON files contain data like plugin/theme/... name, version, svn path and svn revision. They get parsed by the wp4nix `default.nix`, which exposes a bunch of attrsets, so you can use things like `wordpressPackages.plugins.woocommerce`.

The reason why we're using SVN is that with the setup they have there you can republish versions and people do. So the zip you download today does not necessarily have the same hash as the one you download tomorrow.
As a user of it, you'd download all the metadata (so 25MB JSON), but you only build and install the plugins and themes you actually use.

As for this PR:
1) I can't really comment on the Apache2 config changes, because I don't remember when I last configured that software. Although it looks like they're just configuring a virtualhost as opposed to… whatever it's doing now? Just using the default host? What?
2) The update scripts should probably moved up a folder, because you only need one for all themes and one for all plugins, right? You'd need to pass the name, but you can probably just do that with `writeScript` or something like that. That and they're "flawed" for the reasons outlined above, but you might get away with this approach here, since you would always substitute from the binary cache, once it's in there, anways.
3) The database changes seem sensible enough, but they definitely need some kind of migration instructions. I know WordPress doesn't like if you change that kind of thing, so this needs to be tested well.
4) You might run into issues with setting `WP_HOME` and `WP_SITEURL` the way you are doing here, if you ever change them. The reason for this is because WordPress (themes and plugins, idk if core as well) manage to take this and put it into the database. Which is why there are plugins like [this](https://wordpress.org/plugins/wp-migrate-db/).
5) Does the `nixos/tests` test still pass without any changes?

That said, this still looks like an overall improvement :+1:

mohe2015 committed:
nixos/wordpress: Add option to allow mutable wp-content directory

mohe2015 committed:
nixos/wordpress: Use separate database for each wordpress instance

mohe2015 committed:
nixos/wordpress: Mention a theme and plugin packaging tool in the docs.

mohe2015 commented:
@ajs124 Thanks a lot for the review.

1. I'm assuming you are referring to this commit: https://github.com/NixOS/nixpkgs/commit/39ed207a0c043fbd2b915d86bbbac1e4ad8c34ef?w=1. As far as I understand nix it only changes the service name from the services.wordpress.$name to the current host name the wordpress instance is reachable under, prepended with "wordpress-". This prevents service name clashes if you want to move another service to the old host name.
2. I decided to remove the plugins and themes completely in favor of your approach.
3. I added instructions to the theme path change.
4. I think wordpress core stores the old url in the database but these variables always override the value. But you are right that some/most plugins and themes are incompatible. I removed setting WP_HOME and WP_SITEURL but still added instructions on how to change the hostName as these seem useful.
5. yes

I added an option to allow a mutable wp-content directory but I don't know if this is wanted.

I also added a mention to wp4nix if that is fine with you.

I also added a basic phpmyadmin implementation. Maybe I should put it into another pull request?

Honestly I assume that almost nobody seriously uses this service unmodified because theme updating is currently impossible without loosing settings.

P.S. It would be great if you could add a license to https://git.helsinki.tools/helsinki-systems/wp4nix. Also hosting on github may be beneficial so contributions are possible (If you want to deal with them).
I got a panic when running wp4nix with removed *.json and *.log files for wordpress 5.5: https://gist.github.com/mohe2015/27fe1bf6db92f3ad6a3a72fba8f596c8 If you want to you can look into it.

I would appreciate if you could do a second round of reviewing.

Also I don't know how the stateVersion thing should be handled as I don't want to break unstable users' systems but it seems like this is the current way to go. I hope they are watching the repository closely. (I would need to change the stateVersion changes to 20.09 if this actually manages to get merged in the current release cycle, right?)

ajs124 commented on pkgs/servers/web-apps/wordpress/wordpress-core.nix:
```diff
@@ -0,0 +1,12 @@
+{ stdenv, wordpress }:
+
+with stdenv.lib;
```
What is this needed for?

mohe2015 replied:
You mean the whole file? Wordpress ships with some themes (twentytwenty currently I think, maybe also twentynineteen and twentyeightteen) and plugins (akismet, hello dolly) and these can't be removed or updated if there isn't a new wordpress version released yet. So by removing them you can independently remove or update them using the plugins and themes attribute and therefore also limit the amount of potentially vulnerable code.

mohe2015 replied:
If you meant the stdenv thingy - I removed it.

ajs124 commented:
> As far as I understand nix it only changes the service name from the services.wordpress.$name to the current host name the wordpress instance is reachable under, prepended with "wordpress-". This prevents service name clashes if you want to move another service to the old host name.

Goes to show how much I know about the apache nixos module.

> I added an option to allow a mutable wp-content directory but I don't know if this is wanted. I usually use this for trying out new themes and plugins.

AFAIR mutable content is often needed, because that's also where uploaded images and other files are stored.

> I also added a basic phpmyadmin implementation. Maybe I should put it into another pull request?

Yeah, that should probably go in a separate PR.

> Honestly I assume that almost nobody seriously uses this service unmodified because theme updating is currently impossible without loosing settings.

I kind of doubt anybody uses this, as well.

wp4nix is now licensed under the EUPL-1.2, but given that all code is either by me or @SlothOfAnarchy, we can also re-license it if anyone would prefer to have it under any other license.
Moving it to github is possible but not trivial, as we use gitlab ci and it's integrated into our deployment.

>  I got a panic when running wp4nix with removed *.json and *.log files for wordpress 5.5: https://gist.github.com/mohe2015/27fe1bf6db92f3ad6a3a72fba8f596c8 If you want to you can look into it.

If you want to regenerate the json files from scratch, you probably need to create empty (aka containing `{}`) json files, as it barely does any error handling.
Seeing as this was quite literally my first go project, I'm surprised it didn't set your computer on fire ^^
As far as I remember, it also doesn't actually do any version checking with the wordpress version you give it, because… well, because it wasn't necessary for the plugins and themes we use, probably. Maybe it gets used for the languages? We pull those from subversion as well, though.

> Also I don't know how the stateVersion thing should be handled as I don't want to break unstable users' systems but it seems like this is the current way to go. I hope they are watching the repository closely. (I would need to change the stateVersion changes to 20.09 if this actually manages to get merged in the current release cycle, right?)

I haven't used `stateVersion` so far, maybe @aanderse has some insight on that.

aanderse commented:
You really would think no one uses this module... but I was shocked to learn that a number of people used the `httpd` submodules options before I deprecated/removed them. It goes to show you really don't know :man_shrugging:

I only wrote this module to get rid of the `httpd` submodule options (hint: these options were bad) and don't have any personal interest in maintaining/using this module. I know `nginx` is much more popular compared to `httpd` (especially in NixOS) so maybe it makes sense to do a hard break and rewrite the module for `nginx` the way you would like it, if you're willing to maintain this module in NixOS/nixpkgs... what do you think?

mohe2015 commented:
@aanderse The changes itself would be backwards compatible if the thing with the stateVersion is discussed.

I'm sorry but I won't rewrite this module. Also the better way would be to get a webserver-agnostic interface but I know it's gonna take a while until that's going to happen.

Offtopic:
Also I'm not using this any more because of https://github.com/NixOS/nixpkgs/issues/20836#issuecomment-707991243 and legal issues in Germany with running a server (to me it's not always clear who is responsible if you server got hacked and used to attack somebody else or so). I'm only running four wordpress sites and a nextcloud server so I'm using debian because I think I could argue in court that using debian is pretty standard and they also have a big security team and therefore you can't be hold liable for the damage as long as you updated all the time.

I would still like to get this merged though. The only issue left is what do do with the stateVersion because I don't want to break the system of unstable users. How should I handle that?

mohe2015 commented:
I have rebased and dropped some commits that are not necessary in my opinion. I haven't tested again though yet.

The thing that really is necessary is https://github.com/NixOS/nixpkgs/pull/96910/commits/319e9b2e74fef8a7c22a5fb595ea331168d965ad. But the current migration is not a good way to manage this as this could prevent people from upgrading if they don't want to make these changes. I would propose adding some option like `wordpress.service.usePluginAndThemeDirectoriesWIthoutVersion = true;` or so. What do you think?

mohe2015 commented:
@GrahamcOfBorg test wordpress

mohe2015 commented:
@GrahamcOfBorg test wordpress

mohe2015 commented:
@onny Would love to get a review from you.

I removed the controversial change for now. I will try this in a separate PR.

Other reviews are also greatly appreciated.

mohe2015 commented on nixos/modules/services/databases/mysql.nix:
```diff
@@ -483,7 +483,7 @@ in
               ( echo "CREATE USER IF NOT EXISTS '${user.name}'@'localhost' IDENTIFIED WITH ${if isMariaDB then "unix_socket" else "auth_socket"};"
                 ${concatStringsSep "\n" (mapAttrsToList (database: permission: ''
-                  echo "GRANT ${permission} ON ${database} TO '${user.name}'@'localhost';"
+                  echo 'GRANT ${permission} ON ${database} TO `${user.name}`@`localhost`;'
```
I checked (code-wise) that this shouldn't break anything but somebody else should probably also check that. Also I would run some other modules that use this before merging.

mohe2015 commented on nixos/modules/services/web-apps/wordpress.nix:
```diff
@@ -168,7 +191,7 @@ let
           name = mkOption {
             type = types.str;
-            default = "wordpress";
+            default = if lib.versionAtLeast config.system.stateVersion "21.11" then "wordpress-${name}" else "wordpress";
```
Maybe I should change this to
```suggestion
            default = if lib.versionAtLeast config.system.stateVersion "21.11" then "wordpress-${replaceStrings [ "." ] [ "_" ] name}" else "wordpress";
```

mohe2015 commented:
Superseded by https://github.com/NixOS/nixpkgs/pull/135966. Thank your to everyone who participated.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/45
